src/util/buf.rs

```diff
@@ -11,93 +11,418 @@
+    #[inline]
+    pub fn is_full(&self) -> bool {
+        self.len() == self.capacity()
     }
 
+    #[inline]
     pub fn is_empty(&self) -> bool {
         self.len() == 0
     }
+
+    /// Get the written buf.
+    fn slice(&self) -> (&[u8], &[u8]) {
+        unsafe {
+            let buf = self.buf_as_slice();
+            if self.end >= self.start {
+                (&buf[self.start..self.end], &[])
+            } else {
+                (&buf[self.start..], &buf[..self.end])
+            }
+        }
+    }
+
+    /// Get the not written buf.
+    fn slice_append(&mut self) -> (&mut [u8], &mut [u8]) {
+        if self.is_full() {
+            return (&mut [], &mut []);
+        }
+        unsafe {
+            let start = self.start;
+            let end = self.end;
+            let cap = self.capacity();
+            let buf = self.buf_as_slice_mut();
+
+            if start == 0 {
+                (&mut buf[end..cap], &mut [])
+            } else if start <= end {
+                let (right, left) = buf.split_at_mut(end);
+                let (right, _) = right.split_at_mut(start - 1);
+                (left, right)
+            } else {
+                (&mut buf[end..start - 1], &mut [])
+            }
+        }
+    }
+
+    /// Ensure the capacity of inner buf not less than `capacity`.
+    ///
+    /// If capacity is larger than inner buf, a larger buffer will be reallocated.
+    /// Allocated buffer's capacity doesn't have to be equal to specified value.
+    pub fn ensure(&mut self, capacity: usize) {
+        if capacity <= self.capacity() {
+            return;
+        }
+
+        let cap = self.buf.cap();
+        self.buf.reserve(cap, capacity + 1 - cap);
+        let new_cap = self.buf.cap();
+
+        unsafe {
+            if self.start <= self.end {
+                // written data are linear, no need to move.
+                return;
+            } else if new_cap - cap > self.end && self.end <= cap - self.start {
+                // left part can be fit in new buf and is shorter.
+                let left = self.buf.ptr();
+                let new_pos = self.buf.ptr().offset(cap as isize);
+                ptr::copy_nonoverlapping(left, new_pos, self.end);
+                self.end += cap;
+            } else {
+                let right = self.buf.ptr().offset(self.start as isize);
+                self.start = new_cap - (cap - self.start);
+                let new_pos = self.buf.ptr().offset(self.start as isize);
+                if self.start >= cap {
+                    ptr::copy_nonoverlapping(right, new_pos, new_cap - self.start);
+                } else {
+                    ptr::copy(right, new_pos, new_cap - self.start);
+                }
+            }
+        }
+    }
+
+    /// shrink the inner buf to specified capacity.
+    pub fn shrink_to(&mut self, len: usize) {
+        assert!(len < self.capacity());
+
+        let new_cap = len + 1;
+        let to_keep = cmp::min(len, self.len());
```

Comment:
I don't think *Since shrink_to will drop some data, then the data remain is useless* is always true.